package-tests.cake

```diff
@@ -0,0 +1,312 @@
+        CreatePackageInstallDirectory();
+
+        RunPackageTests();
+    }
+
+    // Default is to just unzip... individual package testers may override
+    protected virtual void CreatePackageInstallDirectory()
+    {
+        _context.CleanDirectory(PackageInstallDirectory);
+        _context.Unzip(PackageUnderTest, PackageInstallDirectory);
+    }
+
+    private void RunPackageTests()
+    {
+        var reporter = new ResultReporter(PackageName);
+
+        foreach (var packageTest in PackageTests)
+        {
+            var resultFile = _outputDir + "TestResult.xml";
+            // Delete result file ahead of time so we don't mistakenly
+            // read a left-over file from another test run. Leave the
+            // file after the run in case we need it to debug a failure.
+            if (_context.FileExists(resultFile))
+                _context.DeleteFile(resultFile);
+
+            DisplayBanner(packageTest.Description);
+
+            int rc = _context.StartProcess(
+                PackageBinDir + "nunit3-console.exe",
+                new ProcessSettings()
+                {
+                    Arguments = packageTest.Arguments,
+                    WorkingDirectory = _outputDir
+                });
+
+            try
+            {
+                var result = new ActualResult(resultFile);
+                var report = new TestReport(packageTest, result);
+                reporter.AddReport(report);
+
+                Console.WriteLine(report.Errors.Count == 0
+                    ? "\nSUCCESS: Test Result matches expected result!"
+                    : "\nERROR: Test Result not as expected!");
+            }
+            catch (Exception ex)
+            {
+                reporter.AddReport(new TestReport(packageTest, ex));
+
+                Console.WriteLine("\nERROR: No result found!");
+            }
+        }
+
+        bool hadErrors = reporter.ReportResults();
+        Console.WriteLine();
+
+        if (hadErrors)
+            throw new Exception("One or more package tests had errors!");
+    }
+
+    private void DisplayBanner(string message)
+    {
+        Console.WriteLine("\n=================================================="); ;
+        Console.WriteLine(message);
+        Console.WriteLine("==================================================");
+    }
+}
+
+public abstract class NetFXPackageTester : PackageTester
+{
+    public NetFXPackageTester(ICakeContext context, string packageVersion)
+        : base(context, packageVersion)
+    {
+        // Add common tests for running under .NET Framework
+        PackageTests.Add(new PackageTest(
+            "Run mock-assembly.dll under .NET 3.5",
+            "net35/mock-assembly.dll",
+            new ExpectedResult("Failed")
+            {
+                Total = 37,
+                Passed = 23,
+                Failed = 5,
+                Warnings = 0,
+                Inconclusive = 1,
+                Skipped = 7
+            }));
+
```

Comment:
Suggestion - how about extracting these to static objects called "StandardResult" and "TwoAssemblyResult" or something, so it's only in one place? Just thinking it could become a lot to update whenever a new test is added to mock assembly, if these package tests grow more.

Reply:
We already have more than two sets. Each derived tester can add tests. The msi and zip testers add more test cases which the nuget and chocolatey runs cannot handle.

I agree with you  that it would be nicer to have all these in one place or at least in one file but I couldn't think of a clean way
to do that. Basically, it seems to require two matching hierarchies. That calls for the Bridge pattern, which seems like a bit of overkill for what is basically a script.

Again, I'd prefer to let this idea percolate rather than trying to solve it all at once. If you like, we can create "idea" issues for some of this stuff.